File: BillPayments/Core/Async.swift
enum Async<T> {
    case loading(previous: T?)
    case data(T)
    case error(any Error)

    var valueOrNull: T? {
        switch self {
        case .data(let value): value
        case .loading(let previous): previous
        case .error: nil
        }
    }
}

File: app/src/main/java/com/billpayments/core/async/Async.kt
package com.billpayments.core.async

sealed interface Async<out T> {
    data object Loading : Async<Nothing>
    data class Data<out T>(val value: T) : Async<T>
    data class Error(val error: Throwable) : Async<Nothing>
}

val <T> Async<T>.valueOrNull: T? get() = (this as? Async.Data<T>)?.value
